# Importing Libraries

In [36]:
import numpy as np
import networkx as nx
import matplotlib.pyplot as plt

# Task 1: Creating a Directed Graph 

In [37]:
graph = nx.read_edgelist("connections.txt", create_using=nx.DiGraph())

# Task 2: Showing Bridge Nodes

In [18]:
# Convert to undirected graph
graph_undirected = graph.to_undirected()
bridges = list(nx.bridges(graph_undirected))
print(bridges)

[('0', '11'), ('0', '12'), ('0', '15'), ('0', '18'), ('0', '37'), ('0', '43'), ('0', '74'), ('0', '114'), ('0', '209'), ('0', '210'), ('0', '215'), ('0', '287'), ('0', '292'), ('0', '335'), ('107', '911'), ('107', '918'), ('107', '1096'), ('107', '1119'), ('107', '1145'), ('107', '1206'), ('107', '1386'), ('107', '1466'), ('107', '1560'), ('107', '1581'), ('107', '1834'), ('348', '358'), ('348', '447'), ('348', '550'), ('414', '585'), ('414', '602'), ('414', '607'), ('414', '608'), ('414', '613'), ('414', '624'), ('414', '638'), ('414', '668'), ('414', '674'), ('1684', '2842'), ('1684', '3031'), ('1684', '3071'), ('1684', '3183'), ('1684', '3230'), ('1912', '2079'), ('1912', '2195'), ('1912', '2269'), ('1912', '2457'), ('1912', '2470'), ('1912', '2569'), ('1912', '2596'), ('3437', '3451'), ('3437', '3453'), ('3437', '3570'), ('3437', '3650'), ('3437', '3709'), ('3437', '3729'), ('3437', '3748'), ('3437', '3798'), ('3437', '3820'), ('3437', '3853'), ('3437', '3856'), ('3437', '3935'), (

# Task 3: Showing Density of the Graph

In [16]:
print(f'Graph Density: {nx.density(graph)}')

Graph Density: 0.0054099817517196435


The density of a graph is a measure that indicates the proportion of potential edges in a graph that are actually present. It is a value between 0 and 1, where 0 represents a completely sparse graph (no edges) and 1 represents a fully connected graph (all possible edges are present).  
The density of the graph is approximately 0.0054. This value is relatively low, indicating that the graph is sparse and has relatively few edges compared to the total number of possible edges. In other words, the low density suggests that there are many nodes with few connections. 

# Task 4: Showing nodes with highest and lowest number of connections

In [25]:
# Get the degree of each node in the graph
node_degrees = graph.degree()

# Find the maximum and minimum degrees in the graph
max_degree = max(node_degrees, key=lambda x: x[1])[1]
min_degree = min(node_degrees, key=lambda x: x[1])[1]

# Find nodes with the maximum and minimum degrees
nodes_with_max_degree = [node for node, degree in node_degrees if degree == max_degree]
nodes_with_min_degree = [node for node, degree in node_degrees if degree == min_degree]

print(f'Maximum Degree: {max_degree}')
print(f'Nodes with highest number of connections: {nodes_with_max_degree}')
print('')
print(f'Minimum Degree: {min_degree}')
print(f'Nodes with lowest number of connections: {nodes_with_min_degree}')

Maximum Degree: 1045
Nodes with highest number of connections: ['107']

Minimum Degree: 1
Nodes with lowest number of connections: ['11', '12', '15', '18', '37', '43', '74', '114', '209', '210', '215', '287', '292', '335', '911', '918', '1096', '1119', '1145', '1206', '1386', '1466', '1560', '1581', '1834', '358', '447', '550', '585', '602', '607', '608', '613', '624', '638', '668', '674', '692', '801', '875', '883', '891', '892', '2842', '3031', '3071', '3183', '3230', '2079', '2195', '2269', '2457', '2470', '2569', '2596', '3451', '3453', '3570', '3650', '3709', '3729', '3748', '3798', '3820', '3853', '3856', '3935', '3974', '3984', '4008', '4010', '4015', '4022', '4024', '4035']


# Task 5: Showing which nodes have the highest incoming and outgoing connections.

In [28]:
# Get the in-degree and out-degree of each node in the graph
in_degrees = graph.in_degree()
out_degrees = graph.out_degree()

# Find the maximum and minimum in-degrees and out-degrees in the graph
max_in_degree = max(in_degrees, key=lambda x: x[1])[1]
max_out_degree = max(out_degrees, key=lambda x: x[1])[1]

# Find nodes with the maximum and minimum in-degrees and out-degrees
nodes_with_max_in_degree = [node for node, degree in in_degrees if degree == max_in_degree]
nodes_with_max_out_degree = [node for node, degree in out_degrees if degree == max_out_degree]

print(f'Maximum In-Degree: {max_in_degree}')
print(f'Nodes with highest number of connections: {nodes_with_max_in_degree}')
print('')
print(f'Minimum Out-Degree: {max_out_degree}')
print(f'Nodes with lowest number of connections: {nodes_with_max_out_degree}')

Maximum In-Degree: 251
Nodes with highest number of connections: ['1888']

Minimum Out-Degree: 1043
Nodes with lowest number of connections: ['107']


# Task 6: Showing nodes with highest closeness, betweenness, and eigenvector. 

In [33]:
# Measure Closeness Centrality
closeness_centrality = nx.closeness_centrality(graph)
max_closeness_value = max(closeness_centrality.values())
nodes_with_max_closeness = [node for node, value in closeness_centrality.items() if value == max_closeness_value]

# Measure Betweenness Centrality
betweenness_centrality = nx.betweenness_centrality(graph)
max_betweenness_value = max(betweenness_centrality.values())
nodes_with_max_betweenness = [node for node, value in betweenness_centrality.items() if value == max_betweenness_value]

# Measure Eigenvector Centrality
eigenvector_centrality = nx.eigenvector_centrality(graph, max_iter = 500)
max_eigenvector_value = max(eigenvector_centrality.values())
nodes_with_max_eigenvector = [node for node, value in eigenvector_centrality.items() if value == max_eigenvector_value]

# Print the results
print("Nodes with Maximum Closeness Centrality Value (", max_closeness_value, "):", nodes_with_max_closeness)
print("==============================================================================")
print("Nodes with Maximum Betweenness Centrality Value (", max_betweenness_value, "):", nodes_with_max_betweenness)
print("==============================================================================")
print("Nodes with Maximum Eigenvector Centrality Value (", max_eigenvector_value, "):", nodes_with_max_eigenvector)


Nodes with Maximum Closeness Centrality Value ( 0.11797503012816456 ): ['2642']
Nodes with Maximum Betweenness Centrality Value ( 0.03299985344063623 ): ['1684']
Nodes with Maximum Eigenvector Centrality Value ( 0.8027042996706522 ): ['2655']


#### Closeness Centrality:
* Node 2642 has the maximum closeness centrality value of approximately 0.118.
* *Interpretation:* Node 2642 is particularly central in terms of closeness, indicating that it is relatively close to other nodes in the network. This could imply efficient information flow or influence spread from Node 2642 to other nodes.
#### Betweenness Centrality:
* Node 1684 has the maximum betweenness centrality value of approximately 0.033.
* *Interpretation:* Node 1684 plays a crucial role in the network's communication flow. It acts as a key intermediary along many shortest paths between other nodes. Removing or influencing Node 1684 may significantly impact information flow in the network.
#### Eigenvector Centrality:
* Node 2655 has the maximum eigenvector centrality value of approximately 0.803.
* *Interpretation:* Node 2655 is highly influential in the network according to eigenvector centrality. This measure considers not only direct connections but also the importance of nodes connected to Node 2655. Node 2655 is highly valuable in reaching other influential nodes in the network.

# Task 7: Implementing a Community Detection Algorithm on the Directed Graph 

In [45]:
# Use Label Propagation Algorithm for community detection
communities = list(nx.algorithms.community.greedy_modularity_communities(graph))

# Get the number of communities
num_communities = len(communities)

# Print the number of communities
print("Number of Communities (Greedy Modularity Algorithm):", num_communities)


Number of Communities (Greedy Modularity Algorithm): 18


# Task 8: Showing the largest and the smallest community

In [46]:
largest_community = max(communities, key=len)
smallest_community = min(communities, key=len)
print(f'Largest community: {len(largest_community)}')
print(f'Smallest community: {len(smallest_community)}')

Largest community: 1053
Smallest community: 3


#### Largest Community:

The largest community consists of 1053 nodes. This suggests that a substantial portion of the network is tightly interconnected. Members within this community likely have numerous connections with each other, potentially indicating a cohesive group, a densely connected subgroup, or a central hub within the network.

#### Smallest Community:

The smallest community comprises only 3 nodes. This indicates a smaller, more isolated group within the network. It could represent a subgroup with specific characteristics or connections, or it might be an outlier with fewer connections compared to the rest of the network. 